include/circt/Dialect/Pipeline/Pipeline.td

```diff
@@ -14,21 +14,145 @@
 
-#ifdef OP_BASE
-#else
-include "mlir/IR/OpBase.td"
-#endif // OP_BASE
 
+include "mlir/IR/OpBase.td"
 include "mlir/IR/SymbolInterfaces.td"
 include "mlir/IR/RegionKindInterface.td"
 include "mlir/Interfaces/CallInterfaces.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/IR/EnumAttr.td"
 
 def Pipeline_Dialect : Dialect {
   let name = "pipeline";
   let cppNamespace = "::circt::pipeline";
 }
 
+def PipelineOp : Op<Pipeline_Dialect, "pipeline", [
+    IsolatedFromAbove,
+    NoSideEffect,
+    SingleBlockImplicitTerminator<"ReturnOp">,
+    RegionKindInterface,
+    HasOnlyGraphRegion 
+  ]> {
+  let summary = "pipeline operation";
+  let description = [{
+    The "pipeline.pipeline" operation represents a retimeable pipeline.
+    The pipeline contains a single block representing a graph region. Pipeline
+    stages are represented by `pipeline.rt.register` operations. Semantics
+    of values crossing register boundaries are defined by lowering passes.
+    The pipeline representation is centered around providing latency insensitive
+    values (valid signals between stages). Such signals can either be removed
+    (in the feedforward, statically scheduled case), used to control stalling
+    (feedback, statically scheduled case) or in conjunction with handshake signals
+    for dynamically scheduled pipelines.
+    A pipelines' latency sensitivity is based on the I/O of the pipeline - if
+    any in- or output port is an ESI channel, all ports are expected to be ESI
+    channels, and the pipeline is considered latency sensitive.
+    The internal representation of the pipeline is agnostic to the latency
+    insensitivity of the I/O. This is by design - allowing us a single source
+    of truth for lowering either latency sensitive or latency insensitive pipelines.
+
+    A typical flow would go like this:
+
+    An untimed datapath is defined:
+    ```
+    pipeline.pipeline(%in0 : i32, %in1 : i32) -> (i32) {
+      ^bb0:(%arg0 : i32, %arg1: i32):
+        %add0 = comb.add %arg0, %arg1 : i32
+        %add1 = comb.add %add0, %arg0 : i32
+        %add2 = comb.add %add1, %arg1 : i32
+        pipeline.return %add2 : i32
+    }
+    ```
+    The datapath is scheduled:
+    ```
+    pipeline.pipeline(%in0 : i32, %in1 : i32) -> (i32) {
+      ^bb0:(%arg0 : i32, %arg1: i32, %go : i1):
+        %add0 = comb.add %arg0, %arg1 : i32
+
+        %s0_valid = pipeline.stage when %go
+        %add1 = comb.add %add0, %arg0 : i32
+  
+        %s1_valid = pipeline.stage when %g1
+        %add2 = comb.add %add1, %arg1 : i32
+
+        pipeline.return %add2 valid %s1_valid : i32
+    }
+    ```
+
+    Stage-crossing dependencies are made explicit through registers.
+    ```
+    pipeline.pipeline(%in0 : i32, %in1 : i32) -> (i32) {
+      ^bb0:(%arg0 : i32, %arg1: i32):
+        %add0 = comb.add %arg0, %arg1 : i32
+
+        %s0_valid, %add0_r = pipeline.stage when %go regs (%add0: i32)
```

Comment:
> Are the registers detected automatically?

Yes - locally, i have a `pipeline-make-regs-explicit` pass which performs a dataflow analysis and mutates the `pipeline.stage` operations accordingly (def-use references across stages).

> Are the registers ... required?

No - this is by design. The lowering pass is fairly dumb in that it pretty much just convers the `pipeline.stage` operation to `seq.compreg` operations. Any dataflow analysis and mutation of stage ops should have done beforehand. The intention here is to keep things separated, that is, analyzing and inferring registers ("hard"-ish) and lowering to HW (easy).

Reply:
I must be missing something... If a cross stage value isn't listed, won't the resulting circt be wrong? Of at the very least have an II >= 2 (since that value could be a multi-cycle signal).

Reply:
> If a cross stage value isn't listed, won't the resulting circt be wrong? Of at the very least have an II >= 2 (since that value could be a multi-cycle signal).

We might have differing view of at what level this pipeline models the operations. At the moment, it strictly serves as a notion of adding registers to evenly break combinational paths in a dataflow graph. There is no notion of initiation interval or multi-cycle signals.

A pipeline
```mlir
pipeline.pipeline ... {
    %0 = ...
    %1 = ...
    %valid = pipeline.stage when %go 
    %2 = comb.add %0, %1
}
```
lowers to
```mlir
%0 = ...
%1 = ...
%valid = comb.reg %go ...
%2 = comb.add %0, %1
```

whereas if the stage register materialization pass has been run on the pipeline:
```mlir
pipeline.pipeline ... {
    %0 = ...
    %1 = ...
    %valid, %0_r, %1_r = pipeline.stage when %go regs %0 %1
    %2 = comb.add %0_r, %1_r
}
```

that lowers to
```mlir
%0 = ...
%1 = ...
%valid = comb.reg %go ...
%0_r = comb.reg %0 ...
%1_r = comb.reg %1 ...
%2 = comb.add %0_r, %1_r
```

Reply:
> We might have differing view of at what level this pipeline models the operations. At the moment, it strictly serves as a notion of adding registers to evenly break combinational paths in a dataflow graph. There is no notion of initiation interval or multi-cycle signals.

In the presence of multi-cycle paths, every data pipeline has a notion of II since with a multi-cycle path, one can logically only show new (meaningful/valid) data every N cycles (wherein N is the length of the multi-cycle path), yes?

The term 'stage' implies -- to me -- something like a 'barrier' wherein all of the signals moving into the next stage must be registered. So in your first example, assuming that `%0` and `%1` are combinationally driven by inputs and `%2` and `%valid` combinationally drives an output, then `%valid` will be output a cycle 'late' by my mental model of a pipeline.

Reply:
> In the presence of multi-cycle paths, every data pipeline has a notion of II since with a multi-cycle path, one can logically only show new (meaningful/valid) data every N cycles (wherein N is the length of the multi-cycle path), yes?

I don't disagree with you here, I just don't want to move the goalpost for what this PR and abstraction intends to capture in its initial version. These are important things to capture but I'd like to get the basic support in tree and working, before tackling more advanced cases. I intend upcoming commits to add support for e.g. `pipeline.delay` which delays a signal across the next N stages, and we could also add a rudimentary `cycles` attribute to arbitrary SSA values to inform register materialization that a path is multi-cycle.

> The term 'stage' implies -- to me -- something like a 'barrier' wherein all of the signals moving into the next stage must be registered. So in your first example, assuming that %0 and %1 are combinationally driven by inputs and %2 and %valid combinationally drives an output, then %valid will be output a cycle 'late' by my mental model of a pipeline.

The intention of the design is to have an abstraction that easily facilitates insertion of new stages/moving stages around (hence the old 'retimeable' name). This is why hardering/materialization of registers is not a required part of the stage operation but instead a pass late in the pass pipeline.  
I see this abstraction as having different modes - one before, and one after register materialization. Your barrier notion matches my view of the abstraction under the assumption that register materialization is run before lowering.

Reply:
(Discussed offline)
Added `pipeline.stage.register` and removed the `regs` capability from `pipeline.stage`. These two operations cannot exist together, with `pipeline.stage` lowering to `pipeline.stage.registers` (see updated example at top).